gnmi_server/server_test.go

```diff
@@ -2742,6 +2743,79 @@ func TestAuthCapabilities(t *testing.T) {
 	}
 }
 
+func TestCPUUtilization(t *testing.T) {
+    mock := gomonkey.ApplyFunc(linuxproc.ReadStat, func(path string) (*linuxproc.Stat, error) {
+        return &linuxproc.Stat{}, nil
+    })
+
+    defer mock.Reset()
+    s := createServer(t, 8081)
+    go runServer(t, s)
+    defer s.s.Stop()
+
+    tests := []struct {
+        desc        string
+	q           client.Query
+	wantNoti    []client.Notification
+	poll        int
+    }{
+        {
+            desc: "poll query for CPU Utilization",
+	    poll: 10,
+	    q: client.Query{
+                Target: "OTHERS",
+		Type:    client.Poll,
+		Queries: []client.Path{{"platform", "cpu"}},
+		TLS:     &tls.Config{InsecureSkipVerify: true},
+	    },
+	    wantNoti: []client.Notification{
+                client.Connected{},
+		client.Sync{},
+	    },
+        },
+    }
+
+    for _, tt := range tests {
+        t.Run(tt.desc, func(t *testing.T) {
+            q := tt.q
+	    q.Addrs = []string{"127.0.0.1:8081"}
+            c := client.New()
+            defer c.Close()
+            var gotNoti []client.Notification
+            q.NotificationHandler = func(n client.Notification) error {
+                if nn, ok := n.(client.Update); ok {
+                    nn.TS = time.Unix(0, 200)
+		    gotNoti = append(gotNoti, nn)
+                } else {
+                    gotNoti = append(gotNoti, n)
+	        }
+                return nil
+            }
+
+            go func() {
+                // wait for stats buffer to contain 3000 obj
+                time.Sleep(time.Second * 300)
+                if err := c.Subscribe(context.Background(), q); err != nil {
+                    t.Errorf("c.Subscribe(): got error %v, expected nil", err)
+                }
+            }()
+
+            // wait for a second after subscribeRequest to sync
+            time.Sleep(time.Second * 301)
```

Comment:
Yes, I agree that it is a long time

The reason I chose 300 is because referring to https://github.com/sonic-net/sonic-gnmi/blob/master/sonic_data_client/non_db_client.go#L349 that is how long it will take to fill the entire stats buffer (3000 elements) since it reads from /proc/stat every 0.1 seconds which is 300 seconds for the entire buffer to fill starting from 0.

I want the stats buffer to be full so I can get full coverage when testing for each scenario inside https://github.com/sonic-net/sonic-gnmi/blob/master/sonic_data_client/non_db_client.go#L142 (100 ms, 1 s, 5s, 1m, 5m)

I tried mocking statsR buffer and pollStats using gomonkey but could not figure out how to do so without significantly changing the existing code.

Reply:
Great!